rustls/src/client/hs.rs

```diff
@@ -419,6 +376,56 @@ fn emit_client_hello_for_retry(
     }
 }
 
+fn prepare_resumption<'a>(
+    resuming: &'a Option<persist::Retrieved<ClientSessionValue>>,
+    exts: &mut Vec<ClientExtension>,
+    doing_retry: bool,
+    suite: Option<SupportedCipherSuite>,
+    cx: &mut ClientContext<'_>,
+    config: &ClientConfig,
+) -> Option<persist::Retrieved<&'a persist::Tls13ClientSessionValue>> {
+    if !config.enable_tickets {
+        return None;
+    }
+
+    // Check whether we're resuming with a non-empty ticket.
+    let resuming = match resuming {
+        Some(resuming) if !resuming.ticket().is_empty() => resuming,
+        _ => {
+            // If we don't have a ticket, request one.
+            exts.push(ClientExtension::SessionTicket(ClientSessionTicket::Request));
+            return None;
+        }
+    };
+
+    let tls13 = match resuming.map(|csv| csv.tls13()) {
+        Some(tls13) if config.supports_version(ProtocolVersion::TLSv1_3) => tls13,
+        None if config.supports_version(ProtocolVersion::TLSv1_2) => {
+            exts.push(ClientExtension::SessionTicket(ClientSessionTicket::Offer(
+                Payload::new(resuming.ticket()),
+            )));
+            return None;
+        }
+        _ => return None,
+    };
```

Comment:
It's not immediately clear from reading this that `None` corresponds to the case where the retrieved session value is for TLS 1.2. In the interest of further linearizing this logic and making it more explicit, how about:

```rust
    if let ClientSessionValue::Tls12(_) = resuming.deref() {
        if config.supports_version(ProtocolVersion::TLSv1_2) {
            exts.push(ClientExtension::SessionTicket(ClientSessionTicket::Offer(
                Payload::new(resuming.ticket()),
            )));
        }
        return None;
    }

    if !config.supports_version(ProtocolVersion::TLSv1_3) {
        return None;
    }

    let Some(tls13) = resuming.map(|csv| csv.tls13()) else {
        return None;
    };
```

Reply:
IMO one of the core values of the linear approach is to check each condition only once. In your suggested code, this is not the case: even though we know after the first `if` that `resuming` is `Tls13`, you need a final `if` to extract the `Retrieved<Tls13ClientSessionValue>` out of the `Retrieved<ClientSessionValue>`, but that branch will always be taken. I tried to add more comments instead to make it clearer what's going on.